spring-cloud-stream-integration-tests/src/test/java/org/springframework/cloud/stream/config/StreamListenerTestMethodWithReturnMessage.java

```diff
@@ -0,0 +1,115 @@
+import org.springframework.cloud.stream.test.binder.MessageCollector;
+import org.springframework.context.ConfigurableApplicationContext;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.handler.annotation.SendTo;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Marius Bogoevici
+ * @author Ilayaperumal Gopinathan
+ */
+@RunWith(Parameterized.class)
+public class StreamListenerTestMethodWithReturnMessage {
+
+	private Class<?> configClass;
+
+	public StreamListenerTestMethodWithReturnMessage(Class<?> configClass) {
+		this.configClass = configClass;
+	}
+
+	@Parameterized.Parameters
+	public static Collection InputConfigs() {
+		return Arrays.asList(new Class[]{TestPojoWithMessageReturn1.class, TestPojoWithMessageReturn2.class});
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	public void testReturnMessage() throws Exception {
+		ConfigurableApplicationContext context = SpringApplication
+				.run(this.configClass, "--server.port=0");
+		MessageCollector collector = context.getBean(MessageCollector.class);
+		Processor processor = context.getBean(Processor.class);
+		String id = UUID.randomUUID().toString();
+		processor.input()
+				.send(MessageBuilder.withPayload("{\"foo\":\"barbar" + id + "\"}")
+						.setHeader("contentType", "application/json").build());
+		TestPojoWithMessageReturn testPojoWithMessageReturn = context
+				.getBean(TestPojoWithMessageReturn.class);
+		assertThat(testPojoWithMessageReturn.receivedPojos).hasSize(1);
+		assertThat(testPojoWithMessageReturn.receivedPojos.get(0)).hasFieldOrPropertyWithValue("foo", "barbar" + id);
+		Message<StreamListenerTestInterfaces.BarPojo> message = (Message<StreamListenerTestInterfaces.BarPojo>) collector
+				.forChannel(processor.output()).poll(1, TimeUnit.SECONDS);
+		assertThat(message).isNotNull();
+		assertThat(message.getPayload().getBar()).isEqualTo("barbar" + id);
+		context.close();
+	}
+
+	@EnableBinding(Processor.class)
+	@EnableAutoConfiguration
+	public static class TestPojoWithMessageReturn1 extends TestPojoWithMessageReturn {
+
+		@StreamListener(Processor.INPUT)
+		@SendTo(Processor.OUTPUT)
+		public Message<?> receive(StreamListenerTestInterfaces.FooPojo fooPojo) {
+			this.receivedPojos.add(fooPojo);
+			StreamListenerTestInterfaces.BarPojo barPojo = new StreamListenerTestInterfaces.BarPojo();
+			barPojo.setBar(fooPojo.getFoo());
+			return MessageBuilder.withPayload(barPojo).setHeader("foo", "bar").build();
+		}
+	}
+
+	@EnableBinding(Processor.class)
+	@EnableAutoConfiguration
+	public static class TestPojoWithMessageReturn2 extends TestPojoWithMessageReturn {
+
+		@StreamListener(Processor.INPUT)
+		@Output(Processor.OUTPUT)
+		public Message<?> receive(StreamListenerTestInterfaces.FooPojo fooPojo) {
+			this.receivedPojos.add(fooPojo);
+			StreamListenerTestInterfaces.BarPojo bazPojo = new StreamListenerTestInterfaces.BarPojo();
```

Comment:
identical in this case as well except for the name of the `bazPojo` var... was this intended to be something else?